datastore/providers/azuresearch_datastore.py

```diff
@@ -0,0 +1,259 @@
+                    FIELDS_EMBEDDING: chunk.embedding,
+                    FIELDS_DOCUMENT_ID: document_id,
+                    FIELDS_SOURCE: chunk.metadata.source,
+                    FIELDS_SOURCE_ID: chunk.metadata.source_id,
+                    FIELDS_URL: chunk.metadata.url,
+                    FIELDS_CREATED_AT: chunk.metadata.created_at,
+                    FIELDS_AUTHOR: chunk.metadata.author,
+                })
+            
+                if len(azdocuments) >= MAX_UPLOAD_BATCH_SIZE:
+                    await upload()
+                    azdocuments = []
+
+        if len(azdocuments) > 0:
+            await upload()
+
+        return ids
+
+    async def delete(self, ids: Optional[List[str]] = None, filter: Optional[DocumentMetadataFilter] = None, delete_all: Optional[bool] = None) -> bool:
+        filter = None if delete_all else self._translate_filter(filter)
+        if delete_all or filter is not None:
+            deleted = set()
+            while True:
+                search_result = await self.client.search(None, filter=filter, top=MAX_DELETE_BATCH_SIZE, include_total_count=True, select=FIELDS_ID)
+                if await search_result.get_count() == 0:
+                    break
+                documents = [{ FIELDS_ID: d[FIELDS_ID] } async for d in search_result if d[FIELDS_ID] not in deleted]
+                if len(documents) > 0:
+                    print(f"Deleting {len(documents)} chunks " + ("using a filter" if filter is not None else "using delete_all"))
+                    del_result = await self.client.delete_documents(documents=documents)
+                    if not all([rr.succeeded for rr in del_result]):
+                        raise Exception("Failed to delete documents")
+                    deleted.update([d[FIELDS_ID] for d in documents])
+                else:
+                    # All repeats, delay a bit to let the index refresh and try again
+                    time.sleep(0.25)
+        
+        if ids is not None and len(ids) > 0:
+            for id in ids:
+                print(f"Deleting chunks for document id {id}")
+                await self.delete(filter=DocumentMetadataFilter(document_id=id))
+
+        return True
+
+    async def _query(self, queries: List[QueryWithEmbedding]) -> List[QueryResult]:
+        """
+        Takes in a list of queries with embeddings and filters and returns a list of query results with matching document chunks and scores.
+        """
+        return await asyncio.gather(*(self._single_query(query) for query in queries))
+    
+    async def _single_query(self, query: QueryWithEmbedding) -> QueryResult:
+        """
+        Takes in a single query and filters and returns a query result with matching document chunks and scores.
+        """
+        filter = self._translate_filter(query.filter) if query.filter is not None else None
+        try:
+            k = query.top_k if filter is None else query.top_k * 2
```

Comment:
why does it need to be * 2 if there is no filter? should that be configurable?

Reply:
It's a simple heuristic to help with cases where we might not find enough matching candidates. I was trying to keep the number of configuration parameters on check. People deploying the plugin on their own can always change the code for something so specific. If you think adding a setting for this is particularly important, I can go back and add it, otherwise I'll leave it as-is for now.

Reply:
fair enough